Complexity_classes_def/P.thy

```diff
@@ -0,0 +1,315 @@
+"is_poly_verif c L \<equiv> is_verif c L \<and> poly_verif_time_bounded c \<and> poly_certif_bounded c"
+
+paragraph \<open>NP definition\<close>
+text \<open>NP is the set of languages having a polyverifier\<close>
+definition NP :: "lang set" where
+"NP \<equiv> {L. \<exists>c. is_poly_verif c L}"
+
+
+
+
+subsection \<open>Sanity checks: reduction to a P/NP problem\<close>
+lemma p_sanity:
+  assumes "D' \<in> P" "poly_reduces D D'"
+  shows "D \<in> P"
+  using assms apply (auto simp add:P_def poly_reduces_def)
+proof 
+  fix g f
+  assume assms:"is_polyreduction f D D'"
+       "poly_time_bounded g" "decides g D'"
+  show "poly_time_bounded (f;;g) \<and> decides (f;;g) D "
+  proof
+    show "poly_time_bounded (f;;g)"
+      using assms is_polyreduction_def reduction_poly by auto
+  next 
+    show "decides (f;;g) D"
+      using assms(1) assms(3) reduction_decision_correct is_polyreduction_def by blast 
+qed
+qed
+
+
+lemma np_sanity:
+  assumes "poly_reduces D D'" "D' \<in> NP"
+  shows "D \<in> NP"
+proof -
+  from assms(1) obtain f
+    where f_def: "is_reduction f D D'" "poly_time_bounded f" "poly_result_bounded f"
+    using is_polyreduction_def poly_reduces_def by auto
+  from assms(2) obtain g
+    where g_def: "is_verif g D'" "poly_verif_time_bounded g" "poly_certif_bounded g"
+    using NP_def is_poly_verif_def by force
+  from f_def(1) have "cons_certif f" by (simp add: is_reduction_def)
+  moreover from g_def(1) have "\<forall>y z. \<exists>r. verif g y z r " by (simp add: is_verif_def)
+  ultimately have "poly_certif_bounded (f;;g)" using f_def(3) g_def(3) 
+      poly_result_bounded_poly_certif_bounded   by blast
+  moreover from f_def(1) g_def(1) have "is_verif (f;;g) D" using reduction_verification_correct by simp
+  moreover from f_def(2) f_def(3) g_def(2) have "poly_verif_time_bounded (f;;g)"
+    using verif_reduction_poly \<open>cons_certif f\<close> by blast
+  ultimately show  ?thesis using NP_def is_poly_verif_def  by auto
+qed
+
+subsection \<open>NP-Hard AND NP-Complete Definitions\<close>
+definition NP_hard :: "lang set" where 
+"NP_hard \<equiv> {L'. \<forall>L\<in>NP. poly_reduces L L'}"
+
+definition NP_complete :: "lang set" where
+"NP_complete \<equiv> NP_hard \<inter> NP"
+
+subsection \<open>Defining NP using P\<close>
+paragraph \<open>Idea AND Intuition\<close>
+text \<open>
+Disclaimer: This is not a proof of P = NP.
+
+Verifiers AND deciders are a lot a like. While within P, it's about the existence of polynomially 
+bounded decider, within NP it's about polynomially bounded verifiers. 
+Hence if we encode the input AND certificate in one value. We could do the work of a verifier 
+using a decider who's able to decode them. 
+\<close>
+
+paragraph \<open>Assumptions\<close>
+text \<open>
+For this definition we will need an encoding/entupling AND decoding/detupling programs that convert
+couples of natural numbers into one natural number AND vice versa.
```

Comment:
In this sense, does "couple" mean the same as "tuple"?

Reply:
yep. isn't a 2-tuple a couple ?